### PR TITLE
keycard pin screen fixes

### LINKED ISSUE #8993: Userpicture overlaps with name on login screen for accounts created with keycard (for small screens)
Discovered during testing of https://github.com/status-im/status-react/pull/8764, decided to be fixed separately

# Problem
photo_2019-09-10 15 00 53

**Device:**  LG V20 (5.7-inch; 2,560x1,440 pixels)

### LINKED ISSUE #9059: Add warning message at the last PUK-code attempt
###  User Story

As a user, I want to see that my card will be blocked and I can't use multiaccount anymore if I entered 3 times wrong PIN code and 4 times wrong PUK-code.

### Implementation
Add warning on last PUK-code attempt with a clear explanation that account will be not usable if you enter invalid PUK-code.

### Acceptance Criteria
Warning is implemented

### PR DESCRIPTION
**Pull request by dmitryn**
fixes #8993
fixes #9059

### Summary

* adjust pin enter screen layout for small screen devices
* fix pin reset / unblock flow
* fix sign with keycard button in wallet
* show warning message when pin blocked and puk being entered
* show pin retry counter
* show free pairing slots on device pairing

status: ready

### CODE REVIEW

**Comment by churik** [comment]
@dmitryn please, describe briefly what you tested and what devices did you use.
Thanks.

**Comment by dmitryn** [comment]
@churik i was adjusting pin screen for AVD virtual device with 480x800 resolution

**Comment by churik** [comment]
@dmitryn failed e2e is not related to your PR.

File: src/status_im/ui/screens/hardwallet/pin/styles.cljs [hunk 1]
```diff
@@ -1,5 +1,6 @@
 (ns status-im.ui.screens.hardwallet.pin.styles
   (:require [status-im.ui.components.colors :as colors]
+            [status-im.utils.styles :as styles]
```
Comment (flexsurfer) [inline src/status_im/ui/screens/hardwallet/pin/styles.cljs]: duplicate

File: src/status_im/ui/screens/signing/views.cljs [hunk 2]
```diff
@@ -145,11 +151,32 @@
...
-(defn- sign-with-keycard-button
+(defn sign-with-keycard-button
   [amount-error gas-error]
-  [button/button {:on-press  #(re-frame/dispatch [:signing.ui/sign-with-keycard-pressed])
-                  :disabled? (or amount-error gas-error)
-                  :label     :t/sign-with-keycard}])
+  (let [disabled? (or amount-error gas-error)]
+    [react/touchable-highlight {:on-press #(when-not disabled?
+                                             (re-frame/dispatch [:signing.ui/sign-with-keycard-pressed]))}
+     [react/view {:background-color   colors/black-light
```
Comment (flexsurfer) [inline src/status_im/ui/screens/signing/views.cljs]: could be moved to styles

**Review by flexsurfer — APPROVED** [review]

**Comment by guylouis** [comment]
hey! I need your help understanding what needs to be done to have this one merged. Just following-up on if #8993 and #9059 can be considered closed

**Comment by andmironov** [comment]
looking into this now

**Comment by andmironov** [comment]
Hey! Here are the issues that i've spotted:

**Step 1**
- “Recover key” should be renamed to “Access key". "Access key" should be a white block aligned to the bottom with a 1px border at the top. Tap on “Access key" should open a bottom sheet. Tap on "..." in the toolbar should open a bottom sheet as well, but with a single option "Get a new key";

Screen Shot 2019-10-21 at 11 40 30

- The bottom sheet itself has extra top and bottom paddings. Please keep in mind that in Figma the bottom sheet has extra spacing for iOS's virtual home button which is not needed on all devices

- The active state of the numpad is missing;
- “2 retries yet” message is shown incorrectly.
- The error is persistent, should flash for a ~5 seconds and disappear;

Screen Shot 2019-10-21 at 11 40 15


**Step 2**
- If the connection on step 2 is interrupted, I see an endless spinner in the popup. 



- The desired behavior of this popup is it to disappear after the connection is re-established. The layout and design of this popup has also slightly changed. I also suggest to show an "Are you sure" prompt after the user cancels;

- If I tap “Cancel” in the popup I see the welcome screen, but instead, I should see the sign-in step 1 again (or If I have several keys, the “Your keys” screen);

Screen Shot 2019-10-21 at 11 53 31

- Are we implementing the "Remember me" aka "Don't ask for keycard all the time"? If so, the layout is a bit different. If we implement biometrics ontop it also has a popup shown after the checkbox is checked;

Screen Shot 2019-10-21 at 11 40 49

Screen Shot 2019-10-21 at 12 09 12


P.S
- We never mention that in order to check if the passcode is correct the user needs to tap the card first. Should we? (No design for that yet, just asking);
- What about these errors?

Screen Shot 2019-10-21 at 11 41 05

- What about Frozen accounts?

Screen Shot 2019-10-21 at 12 03 51

---

Everything I mentioned above is in Figma here https://www.figma.com/file/dEIljL7UPbXgsZUA0Q4qlE5E/Onboarding?node-id=4566%3A12


cc @guylouis

**Review by andmironov — COMMENTED** [review]
Left my review in the comment above

**Comment by andmironov** [comment]
UPD added smaller versions for main screens with specs

Screen Shot 2019-10-22 at 11 30 54

Screen Shot 2019-10-22 at 11 30 45

Figma https://www.figma.com/file/dEIljL7UPbXgsZUA0Q4qlE5E/Onboarding?node-id=5314%3A15

File: src/status_im/hardwallet/core.cljs [hunk 3]
```diff
@@ -1035,11 +1038,12 @@
     (fx/merge cofx
               {:hardwallet/get-application-info {:pairing pairing}
                :db                              (-> db
-                                                    (update-in [:hardwallet :pin] merge {:status      nil
-                                                                                         :enter-step  :original
-                                                                                         :current     [0 0 0 0 0 0]
-                                                                                         :puk         []
-                                                                                         :error-label nil}))}
+                                                    (update-in [:hardwallet :pin] merge {:status       nil
```
Comment (yenda) [inline src/status_im/hardwallet/core.cljs]: remove the alignment in the map so that the line isn't above 80 chars

File: translations/en.json [hunk 4]
```diff
@@ -1108,5 +1108,6 @@
 	"lock-app-with": "Lock app with",
 	"grant-face-id-permissions": "To grant the required Face ID permission, please go to your system settings and make sure that Status > Face ID is selected",
 	"request-feature": "Request a feature",
-    "select-account-dapp": "Select the account you wish to use with Dapps"
+  "select-account-dapp": "Select the account you wish to use with Dapps",
```
Comment (yenda) [inline translations/en.json]: a few spaces missing

**Review by yenda — APPROVED** [review]

**Comment by dmitryn** [comment]
@andmironov thanks for the input! I suggest to create separate issues from the points you mentioned and proceed with merging current PR as it's open for a while.

**Comment by guylouis** [comment]
is @andmironov agreement the only thing missing to merge these changes ?

**Comment by dmitryn** [comment]
@guylouis yes. This PR addresses/fixes issues https://github.com/status-im/status-react/issues/8993 https://github.com/status-im/status-react/pull/9086

What @andmironov is suggesting should be addressed too, but in next PRs.

**Pull request merged**